flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/RexNodeTranslator.scala

```diff
@@ -74,15 +75,52 @@ object RexNodeTranslator {
 
   /**
     * Parses all input expressions to [[UnresolvedAlias]].
-    * And expands star to parent's full project list.
+    * And expands star to parent's full project list and flattens composite types.
     */
-  def expandProjectList(exprs: Seq[Expression], parent: LogicalNode): Seq[NamedExpression] = {
+  def expandProjectList(
+      exprs: Seq[Expression],
+      parent: LogicalNode,
+      tableEnv: TableEnvironment)
+    : Seq[NamedExpression] = {
+
     val projectList = new ListBuffer[NamedExpression]
     exprs.foreach {
+
       case n: UnresolvedFieldReference if n.name == "*" =>
         projectList ++= parent.output.map(UnresolvedAlias(_))
+
+      // flattening can only applied on field references
+      case Flattening(composite) if
```

Comment:
I think this would not be intuitive in case of an expression that returns a Pojo such as a user-defined scalar function. As a user I would expect the result of that expression to be flattened. So I would rather throw an exception telling that this is not supported. Expressions with non-composite return types can be handled as below.